### ISSUE TITLE
Improper Regex::replace behaviour

### ISSUE THREAD

**Issue by ConnorGray** [rust-lang-nursery/regex#291]
The regular express `((IMG|CAM|MG|MB2)_|(DSCN|CIMG))(?P<n>[0-9]+)$` is a match on the text `"CIMG2341"` but expected result of `regex.replace("CIMG2341", "$n")` returns `"2341"` is not found, instead `"CIMG2341"` is returned. The correct behavior is observer when the original regex is replaced with `((MG|MB2)_|(DSCN|CIMG))(?P<n>[0-9]+)$`.

```rust
extern crate regex;

use regex::{Regex};

fn main() {
    let text = "CIMG2341";

    let regex = Regex::new(
        "((IMG|CAM|MG|MB2)_|(DSCN|CIMG))(?P<n>[0-9]+)$").unwrap();

    assert!(regex.is_match(text)); // This works (no panic)

    let renamed = regex.replace(text, "$n");
    assert_eq!(renamed, "2341"); // This panics
}
```

**Comment by mbrubeck** [comment]
This is a regression from 63004908030e126ece87102869c36fb7c68895a5 (#290).

**Comment by BurntSushi** [comment]
@ConnorGray Thanks again for the bug report! This should be fixed in `regex 0.1.80` and `regex-syntax 0.3.9` on crates.io.

**Comment by ConnorGray** [comment]
@BurntSushi Thanks again for the quick fix!
